### PR TITLE
Avoid collisions in typed node namespaces, add missing uppercase in alt nodes, remove ShapeTag from ADT alt members

### LINKED ISSUE #1601: ADT unions used as traits have unescaped names
Similar to #1592.

```smithy
$version: "2"

namespace demo

@smithy4s.meta#adt
@trait
union class {
    u: Struct
}

structure Struct {}

@class(u: {

})
string MyShape
```

```scala
[error] ./demo/MyShape.scala:13:10
[error] an identifier expected, but 'class' found
[error]     demo.class.Struct(),
[error]          ^^^^^
```

This doesn't happen with normal unions.

### PR DESCRIPTION
**Pull request by kubukoz**
Closes #1601 and the issue found [in the comments](https://github.com/disneystreaming/smithy4s/issues/1601#issuecomment-2380834208).

## PR Checklist (not all items are relevant to all PRs)

- [ ] Added unit-tests (for runtime code)
- [x] Added bootstrapped code + smoke tests (when the rendering logic is modified)
- [ ] Added build-plugins integration tests (when reflection loading is required at codegen-time)
- [ ] Added alloy compliance tests (when simpleRestJson protocol behaviour is expanded/updated)
- [ ] Updated dynamic module to match generated-code behaviour
- [ ] Added documentation
- [x] Updated changelog

### CODE REVIEW

File: sampleSpecs/reservednames.smithy [hunk 1]
```diff
@@ -146,3 +151,17 @@ structure Scala3ReservedKeywords {
     export: String
     enum: String
 }
+
+@smithy4s.meta#adt
+@trait
```
Comment (kubukoz) [inline sampleSpecs/reservednames.smithy:156]: **thought:** maybe we should just make this illegal? using trait and adt like that

File: modules/codegen/src/smithy4s/codegen/internals/Renderer.scala [hunk 2]
```diff
@@ -1617,7 +1617,10 @@ private[internals] class Renderer(compilationUnit: CompilationUnit) { self =>
       line"${ref.show}.${altName.capitalize}Case.widen".write
 
     case AltTN(_, _, AltValueTN.ProductAltTN(alt)) =>
-      alt.runDefault.write
+      // The `widen` is necessary in Scala 3.
+      // Without it, there are two schemas (the struct and the union) that match the type
+      // for the conversion to Hints.Binding.
+      line"${alt.runDefault}.widen".write
```
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:1622]: Okay, I don't mind the .widen call to help the compiler, but I think the ADT member's schema should not be `implicit`, ever. Would you mind addressing it in this PR ? 

That's something I find regrettable with regards to traditional implicit derivation solutions in Scala 2 : if a case class is an ADT member, it shouldn't be accompanied with its own typeclass instances, as the semantics of a program change drastically based on whether the type is widened or not.

File: sampleSpecs/reservednames.smithy (hunk 1 above)
Comment (Baccata) [reply sampleSpecs/reservednames.smithy:156]: I don't think it should be illegal, the semantics of these two being completely orthogonal. I do however think that dynamic bindings will just solve all these little gotchas.

File: modules/codegen/src/smithy4s/codegen/internals/Renderer.scala (hunk 2 above)
Comment (kubukoz) [reply modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:1622]: I agree, will give it a shot

Comment (kubukoz) [reply modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:1622]: done, this was apparently some previously-missed case, as we do this in all the other cases (just not 0-field product structs in a union)

Comment (kubukoz) [reply modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:1622]: also removing the widen

Comment (kubukoz) [reply modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:1622]: I rushed too much, apparently the widen is still necessary?

Comment (kubukoz) [reply modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:1622]: Widen is back, for Scala 2 :D

**Review by Baccata — APPROVED** [review]

**Pull request merged**